src/rules/fileNameCasingRule.ts

```diff
@@ -28,6 +28,64 @@ enum Casing {
     SnakeCase = "snake-case",
 }
 
+type RegexConfig = Record<string, Casing>;
+
+type SimpleConfig = Casing;
+
+type Config = SimpleConfig | RegexConfig;
+
+type ValidationResult = Casing | undefined;
+
+type Validator<T extends Config> = (sourceFile: ts.SourceFile, casing: T) => ValidationResult;
+
+const rules = [Casing.CamelCase, Casing.PascalCase, Casing.KebabCase, Casing.SnakeCase];
+
+function isCorrectCasing(fileName: string, casing: Casing): boolean {
+    switch (casing) {
+        case Casing.CamelCase:
+            return isCamelCased(fileName);
+        case Casing.PascalCase:
+            return isPascalCased(fileName);
+        case Casing.KebabCase:
+            return isKebabCased(fileName);
+        case Casing.SnakeCase:
+            return isSnakeCased(fileName);
+    }
+}
+
+const validateWithRegexConfig: Validator<RegexConfig> = (sourceFile, casingConfig) => {
+    const fileName = path.parse(sourceFile.fileName).base;
+    const config = Object.keys(casingConfig).map(key => ({
+        casing: casingConfig[key],
+        regex: RegExp(key),
+    }));
+
+    const match = config.find(c => c.regex.test(fileName));
+
+    if (match === undefined) {
+        return undefined;
+    }
+
+    const normalizedFileName = fileName.replace(match.regex, "");
+
+    return isCorrectCasing(normalizedFileName, match.casing) ? undefined : match.casing;
+};
+
+const validateWithSimpleConfig: Validator<Casing> = (sourceFile, casingConfig) => {
+    const fileName = path.parse(sourceFile.fileName).name;
+    const isValid = isCorrectCasing(fileName, casingConfig);
+
+    return isValid ? undefined : casingConfig;
+};
+
+const validate = (sourceFile: ts.SourceFile, casingConfig: Config): ValidationResult => {
+    const validator =
+        typeof casingConfig === "string" ? validateWithSimpleConfig : validateWithRegexConfig;
+
+    // @ts-ignore https://github.com/Microsoft/TypeScript/issues/7294
+    return validator(sourceFile, casingConfig);
```

Comment:
All right, reasoning for the TS error: it's a lack of type inference in the type system that `validator`, either when explicitly typed as a `Validator<Config>` or as the result of the ternary expression, is not allowed to be called this way. In this line, TypeScript doesn't know that `validator` takes in a simple config if `casingConfig === "string"` or a regexp config otherwise.

To be clear: your code is correct, but TypeScript doesn't have the ability to understand that.

Solution: remove the `validator` variable:

```typescript
return casingConfig === "string"
    ? validateWithSimpleConfig(sourceFile, casingConfig)
    : validateWithRegexConfig(sourceFile, casingConfig);
```

You can look for issues on TypeScript if you'd like to see if this has already been reported. Fun stuff!

Reply:
_(and once you fix that, the TSLint error should go away)_